### PR TITLE
Trackdrive logic

### PR DESCRIPTION
**Pull request by b1n-ch1kn**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimisation

## [required] Description

This adds state conditions to the SLAM, planner, reactive controller, and path following controller. This allows for the transition from discovery to path following, particularly the control handover.
Minor edits to the 'Node initialised' messages published across our stack for some consistency and updated naming conventions.

## [required] Documentation

Updated an old mission handling doc found in `Driverless/Documentation/Software/Infrastructure/Competition Mission Launching`
Taking feedback on the terminal interface too: use with `ros2 run terminal_control controller`

### [optional] Usability concerns or breaking changes?

Controller nodes will not be operational unless 'ready to drive' (R2D) is published by the vehicle supervisor node or our terminal interface.
Pure pursuit (python impl) will not be active unless the initial lap is completed with the new state transition. Or use the termainl interface to increment laps as if it had completed them, enabling path following.

### CODE REVIEW

File: src/navigation/planners/planners/node_ordered_mid_spline.py [hunk 1]
```diff
@@ -120,23 +121,38 @@ def sort_cones(cones, start_index=None, end_index=None):
...
         self.get_logger().debug("Received map")
+        self.current_track = track_msg
+
+    def planning_callback(self):
+        # skip if we haven't completed a lap yet
+        if not self.planning:
+            return
 
-        cones = [cone_with_cov.cone for cone_with_cov in track_msg.cones_with_cov]
+        cones = self.current_track.cones
```
Comment (supex0fan) [inline src/navigation/planners/planners/node_ordered_mid_spline.py:155]: Are you sure this doesn't need to be cone_with_cov's cone?

File: src/operations/mission_controller/mission_controller/node_trackdrive_handler.py [hunk 2]
```diff
@@ -1,59 +1,57 @@
...
+            self.last_lap_time = time.time()
+            self.get_logger().info("Trackdrive mission started")
 
     def pose_callback(self, msg: PoseWithCovarianceStamped):
         # check if car has crossed the finish line (0,0)
         # get distance from 0,0 and increment laps when within a certain threshold
         # and distance is increasing away from 0,0
-        if self.started and msg.pose.pose.position.x < 0.3 and msg.pose.pose.position.y < 0.5:
+        if self.r2d and abs(msg.pose.pose.position.x) < 0.5 and abs(msg.pose.pose.position.y) < 2:
             if time.time() - self.last_lap_time > 20:  # 20 seconds at least between laps
```
Comment (supex0fan) [inline src/operations/mission_controller/mission_controller/node_trackdrive_handler.py:40]: Funny how difficult it is to detect such a simple thing. 

I had an idea that might reduce the reliance on arbitrary values here. Store the previous pose's x value, then check the current pose's x value against it, it the previous was negative and the current is positive, and the y value is within the threshold shown here, then you have a new lap. This removes the need for x position checks against an arbitrary value and the minimum lap time check, if it works that is. Still leaves the y position check but the distance between blue and yellow cones is fairly regulated so I don't really consider it an arbitrary value.

File: src/navigation/planners/planners/node_ordered_mid_spline.py (hunk 1 above)
Comment (b1n-ch1kn) [reply src/navigation/planners/planners/node_ordered_mid_spline.py:155]: the track actually populates both fields now, plain cones and cones with cov - so it's one less loop on the receiving end.

File: src/operations/mission_controller/mission_controller/node_trackdrive_handler.py (hunk 2 above)
Comment (b1n-ch1kn) [reply src/operations/mission_controller/mission_controller/node_trackdrive_handler.py:40]: Yeah that's probably a good idea for X. The Y val was more troublesome if the car ran wide, so +-2m is actually wider than the track

Comment (supex0fan) [reply src/operations/mission_controller/mission_controller/node_trackdrive_handler.py:40]: We could take the actual width of the track by subtracting one orange cone from the other, or we could store the y coordinates of two opposing orange cones and check that the car's y is between them.

Comment (supex0fan) [reply src/operations/mission_controller/mission_controller/node_trackdrive_handler.py:41]: Do you want to try what I mentioned in my previous comment or are you happy with where it is here. Also I think you can remove the minimum lap time check here as when the previous x was negative and the current is positive, the last x is set to the positive x and the condition won't get hit again until next lap.

Comment (b1n-ch1kn) [reply src/operations/mission_controller/mission_controller/node_trackdrive_handler.py:41]: Ah I completely missed your additional comment whoops.
Probably fine as it is, 2m is enough clearance so far. Also, the time is just a failsafe if theres any slight drift in the estimated pose which can slightly jiggle around before the car actually starts moving, sending it back and forth over the line.

Comment (b1n-ch1kn) [reply src/operations/mission_controller/mission_controller/node_trackdrive_handler.py:41]: @supex0fan was that all g for approval btw?

**Review by Damin-G — APPROVED** [review]

**Pull request merged**